replicator/run.py

```diff
@@ -57,121 +56,85 @@ def query_serial(zone, server):
+        if self.age < 60 and self.remote_serial == self.serial:
+            return False
 
+        print('Retrieving zone list ...')  # e.g. zone updates where the NOTIFY went lost
+        self._retrieve()
+        return True
 
-def main():
-    print('Loading local zone serials ...')
-    local_zones = {zone['name'] for zone in pdns_request('get', path='/zones').json() if zone['name'] != catalog_domain}
-    print(f'Done. Number of local zones is {len(local_zones)}.')
 
+def main():
     catalog = Catalog()
-    processed_serial = 0
+    processed_serial = None
 
     while True:
         # Note that there may be AXFRs pending from the previous loop iteration. However, it is still useful to fetch
         # the new catalog right away, because:
         #   - it helps discarding useless intermediate states, e.g. if a domain is quickly deleted and recreated,
         #   - if replication is stuck because the catalog is invalid, a new catalog improves chances of recovery,
         #   - waiting for all tasks to be completed would allow long-running AXFRs to hold up new catalog changes.
-        catalog.update()
+        catalog_refreshed = catalog.update()
 
-        # See if the last parsed catalog is the current one
-        if processed_serial == catalog.serial:
+        # Do nothing if catalog has not changed (no domain additions/deletions) and all serials were compared recently
+        if processed_serial == catalog.serial and not catalog_refreshed:
             print(f'Nothing to do (catalog {processed_serial} with {len(local_zones)} zones up to date).')
-            catalog.check_all_serials()
             sleep(1)
             continue
 
-        # Parse current state of catalog zone and determine slave-side changes
-        catalog_zones = catalog.parse()
+        print('Doing comprehensive serial check ...')
+
+        remote_zones = set(catalog.serials.keys())
+        local_serials = {zone['name']: zone['edited_serial'] for zone in pdns_request('get', path='/zones').json()}
+        local_zones = set(local_serials.keys())
 
         # Compute changes
-        deletions = local_zones - catalog_zones
-        additions = catalog_zones - local_zones
+        additions = remote_zones - local_zones
+        deletions = local_zones - remote_zones
+        modifications = {zone for zone, serial in local_serials.items() if catalog.serials.get(zone, 0) > serial}
+
+        # Apply additions
+        for zone in additions:
+            print(f'Adding zone {zone} ...')
+            pdns_request('post', path='/zones', body={'name': zone, 'kind': 'SLAVE', 'masters': [master_ip]})
+            local_zones.add(zone)
+            print(f'Queueing initial AXFR for zone {zone} ...')
+            pdns_request('put', path='/zones/{}/axfr-retrieve'.format(pdns_id(zone)))
+
+        # Apply modifications
+        for zone in modifications:
+            print(f'Queueing AXFR for stale zone {zone} ...')
```

Comment:
add serial info to debug output

Reply:
Not available within that scope, would require changing the `modifications` data structure. Let's do it at another opportunity.